### PR TITLE
Feature/1578/create page my certificates

### PR DESCRIPTION
**Pull request by anemyro**
## Description

There is one problem point in tests coverage. It's my-certificate.js file. The reason is setCertificates & setCount function calls in useQuery in onCompleted aren't covered with tests (attached below at 3-rd screenshot). I tried to change onCompleted to useEffect, in this case the test coverage in this file became 100% but the problem in rendering Certificate Item component occurred.

#### Screenshots

Page:
Screenshot 2022-01-21 at 18 01 12

Tests:
Screenshot 2022-01-21 at 17 50 49

Problem functions to test - setCertificates & setCount 
Screenshot 2022-01-21 at 15 22 08

### Checklist

- [ x] 🔽 My branch is up-to-date with "development" branch
- [ x] ✅All tests passed locally
- [ x] ✨My changes working with up-to-date admin and back-end part locally, like charm
- [ x] 🔗 Link pull request to issue

### CODE REVIEW

**Review by maarekua — APPROVED** [review]

File: src/containers/my-certificates/filled-certificates/filled-certificates.js [hunk 1]
```diff
@@ -1,49 +1,48 @@
...
+import Modal from '../../../components/modal';
 import routes from '../../../configs/routes';
+import { useStyles } from './filled-certificates.styles';
+import CertificateTable from '../certificate-table';
 
 const FilledCertificates = ({ items }) => {
+  const [modalVisibility, setModalVisibility] = useState(false);
   const styles = useStyles();
   const { t } = useTranslation();
+  const modalHandler = () => {
```
Comment (maarekua) [inline src/containers/my-certificates/filled-certificates/filled-certificates.js:15]: Work with naming, for me it looks like 'openModal', and the next one should be 'closeModal', but you can create one handler that will receive true or false, or work with setModalVisibility directly. Cause it is the only thing that this method is working with.

File: src/images/certificates/EmptyCertificateImg.js [hunk 2]
```diff
@@ -0,0 +1,6 @@
+import emptyCertificateLight from './empty-certificate-light-theme-img.png';
+import emptyCertificateDark from './empty-certificate-dark-theme-img.png';
```
Comment (maarekua) [inline src/images/certificates/EmptyCertificateImg.js:2]: Should be SVG.

File: src/locales/en/certificate.json [hunk 3]
```diff
@@ -1,13 +1,16 @@
...
   "price": "Price",
   "expiration": "Expiration date",
   "status": "Status",
-  "buy": "Buy new certificate",
+  "buy": "Buy certificate",
   "add": "Add certificate",
   "active": "Active",
   "expired": "Expired",
-  "used": "Used"
+  "used": "Used",
```
Comment (maarekua) [inline src/locales/en/certificate.json:13]: Check all translations you are adding, maybe there are some in common. Cause 'copy', 'used', UAH looks like common for me.

File: src/pages/my-certificates/my-certificates.js [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+  const [certificates, setCertificates] = useState([]);
+  const [count, setCount] = useState([]);
+  const styles = useStyles();
+
+  const { data, loading, error } = useQuery(getAllCertificates, {
+    variables: {
+      limit: 5,
+      skip: 0
+    },
+    onCompleted: () => {
```
Comment (maarekua) [inline src/pages/my-certificates/my-certificates.js:20]: You should use 'data' that is coming in onCompleted arguments, add one and use it. In another way, you can have bugs.

File: src/tests/unit/containers/my-certificates/certificate-item/certificate-item.spec.js [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+jest.mock('@material-ui/styles', () => ({
+  ...jest.requireActual('@material-ui/styles'),
+  useTheme: () => ({
+    palette: {
+      type: 'light'
+    }
+  })
+}));
+
+jest.mock('react-i18next', () => ({
```
Comment (maarekua) [inline src/tests/unit/containers/my-certificates/certificate-item/certificate-item.spec.js:22]: Doest we have a global mock for i18next? if not add one.

**Review by maarekua — CHANGES_REQUESTED** [review]

File: src/pages/my-certificates/my-certificates.js (hunk 4 above)
Comment (anemyro) [reply src/pages/my-certificates/my-certificates.js:20]: At the beginning I had this way - onComplete: (data), also so this way in other components. But Sonar scolded for a mistake, that I don`t use "data".  When I removed it, Sonar passed well.   
Ok, I`ll return data to onComplete. If Sonar show mistake I`ll inform you

File: src/containers/my-certificates/cetrificate-item/certificate-item.styles.js [hunk 6]
```diff
@@ -4,11 +4,43 @@ export const useStyles = makeStyles(({ palette }) => ({
...
+    width: '110px',
+    height: '28px',
+    pointerEvents: 'none'
+  }),
+  copyBtn: (isLightTheme) => ({
+    backgroundColor: isLightTheme ? palette.backgroundColor : palette.backgroundColor,
+    border: 'none'
+  }),
+  copyIcon: (isLightTheme) => ({
+    backgroundColor: isLightTheme ? palette.backgroundColor : palette.backgroundColor,
```
[the remark this answers is not in the document]
Comment (anemyro) [reply src/containers/my-certificates/cetrificate-item/certificate-item.styles.js:29]: Fixed

File: src/containers/my-certificates/cetrificate-item/certificate-item.js [hunk 7]
```diff
@@ -1,39 +1,61 @@
...
-import CertificateCodeCopy from '../../../images/certificateCodeCopy';
-import Images from './images';
+import CertificateCodeCopy from '../../../images/certificates/certificateCodeCopy';
+import CertificateImages from '../../../images/certificates/CertificateImages';
 
 const CertificateItem = ({ item }) => {
   const styles = useStyles();
   const { t } = useTranslation();
 
+  let status = item.isActive ? 'active' : 'used';
```
[the remark this answers is not in the document]
Comment (anemyro) [reply src/containers/my-certificates/cetrificate-item/certificate-item.js:13]: Status variable can take 3 values: active, used, expired. How to make ternary for 3 values?

File: src/containers/my-certificates/cetrificate-item/certificate-item.js [hunk 8]
```diff
@@ -1,39 +1,61 @@
...
+        <div className={styles.date}>{`${dateHandler(item.dateStart)} - ${dateHandler(
+          item.dateEnd
+        )}`}</div>
       </TableCell>
       <TableCell data-cy='certificate-item-status'>
-        <div className={styles.status}>{t(`certificate.${item.status}`)}</div>
+        {status === 'active' && (
+          <div className={styles.statusGreen}>{t(`certificate.${status}`)}</div>
+        )}
+        {!item.isUsed && !item.isActive && (
```
[the remark this answers is not in the document]
Comment (anemyro) [reply src/containers/my-certificates/cetrificate-item/certificate-item.js:55]: Fixed

File: src/images/certificates/EmptyCertificateImg.js (hunk 2 above)
Comment (anemyro) [reply src/images/certificates/EmptyCertificateImg.js:2]: Fixed

File: src/locales/en/certificate.json (hunk 3 above)
Comment (anemyro) [reply src/locales/en/certificate.json:13]: I checked them. 'Copy' and 'used' are unique. UAH is able in another translation file but it is only one value but another isn't cause in the 'currency' key I have pare: грн. / UAH. So, for my case to use one translation from one file (certificate) and one from another (for example, checkout) will lead to more logic in JSX

File: src/tests/unit/containers/my-certificates/certificate-item/certificate-item.spec.js (hunk 5 above)
Comment (anemyro) [reply src/tests/unit/containers/my-certificates/certificate-item/certificate-item.spec.js:22]: You are right, there is a global mock. I deleted  it

File: src/containers/my-certificates/filled-certificates/filled-certificates.js (hunk 1 above)
Comment (anemyro) [reply src/containers/my-certificates/filled-certificates/filled-certificates.js:15]: Fixed. Change for: 

 const openModal = () => {
    !modalVisibility ? setModalVisibility(true) : setModalVisibility(false);
  };

File: src/containers/my-certificates/certificate-table/certificate-table.js [hunk 9]
```diff
@@ -1,20 +1,27 @@
...
   const { t } = useTranslation();
 
-  const rowItems = ['certificate', 'code', 'price', 'expiration', 'status'];
-  const headerItems = rowItems.map((item) => (
-    <TableCell key={item.id}>{t(`certificate.${item}`)}</TableCell>
+  useEffect(() => {
+    if (items.length) setItemsList(items);
+  }, [items]);
+
+  itemsList.sort((a, b) => new Date(b.dateStart) - new Date(a.dateStart));
```
[the remark this answers is not in the document]
Comment (anemyro) [reply src/containers/my-certificates/certificate-table/certificate-table.js:18]: Fixed

File: src/containers/my-certificates/certificate-table/certificate-table.js [hunk 10]
```diff
@@ -1,20 +1,27 @@
...
 const CertificateTable = ({ items }) => {
+  const [itemsList, setItemsList] = useState([]);
   const styles = useStyles();
   const { t } = useTranslation();
 
-  const rowItems = ['certificate', 'code', 'price', 'expiration', 'status'];
-  const headerItems = rowItems.map((item) => (
-    <TableCell key={item.id}>{t(`certificate.${item}`)}</TableCell>
+  useEffect(() => {
+    if (items.length) setItemsList(items);
```
[the remark this answers is not in the document]
Comment (anemyro) [reply src/containers/my-certificates/certificate-table/certificate-table.js:15]: Fixed

File: src/containers/my-certificates/certificate-table/certificate-table.js [hunk 11]
```diff
@@ -1,20 +1,28 @@
...
 import { useTranslation } from 'react-i18next';
+
 import { Table, TableBody, TableCell, TableHead, TableRow } from '@material-ui/core';
 import { useStyles } from './certificate-table.styles';
-
+import { ROW_FIELDS } from '../../../configs/index';
 import CertificateItem from '../cetrificate-item';
 
 const CertificateTable = ({ items }) => {
+  const certificatesArr = items;
```
Comment (maarekua) [inline src/containers/my-certificates/certificate-table/certificate-table.js:10]: You are making not a copy but a link for items. And still mutating in memorizedItems. You don't need this const. You can make all these logic in useMemo.

File: src/containers/my-certificates/certificate-table/certificate-table.js [hunk 12]
```diff
@@ -1,20 +1,28 @@
...
 
-  const rowItems = ['certificate', 'code', 'price', 'expiration', 'status'];
-  const headerItems = rowItems.map((item) => (
-    <TableCell key={item.id}>{t(`certificate.${item}`)}</TableCell>
+  const memoizedItems = useMemo(
+    () => certificatesArr.sort((a, b) => new Date(b.dateStart) - new Date(a.dateStart)),
+    [certificatesArr]
+  );
+
+  const certificateItems = memoizedItems.map((item) => (
```
Comment (maarekua) [inline src/containers/my-certificates/certificate-table/certificate-table.js:19]: Put it in useMemo.

**Review by maarekua — CHANGES_REQUESTED** [review]

File: src/containers/my-certificates/certificate-table/certificate-table.js (hunk 11 above)
Comment (anemyro) [reply src/containers/my-certificates/certificate-table/certificate-table.js:10]: Can I do such way ? 


—————BEFORE———— 

const CertificateTable = ({ items }) => {
  const certificatesArr = items;
  const styles = useStyles();
  const { t } = useTranslation();

  const memoizedItems = useMemo(
    () => certificatesArr.sort((a, b) => new Date(b.dateStart) - new Date(a.dateStart)),
    [certificatesArr]
  );

  const certificateItems = memoizedItems.map((item) => (
    <CertificateItem item={item} key={item._id} />
  ));

  const headerItems = ROW_FIELDS.map((item) => (
    <TableCell key={item._id}>{t(`certificate.${item}`)}</TableCell>
  ));

  return (…);
};

export default CertificateTable;

—————AFTER———— 

const CertificateTable = ({ items }) => {
  const styles = useStyles();
  const { t } = useTranslation();

  const certificateItems = useMemo(() => {
    return items
      .sort((a, b) => new Date(b.dateStart) - new Date(a.dateStart))
      .map((item) => <CertificateItem item={item} key={item._id} />);
  }, [items]);

  const headerItems = useMemo(() => {
    return ROW_FIELDS.map((item) => (
      <TableCell key={item._id}>{t(`certificate.${item}`)}</TableCell>
    ));
  }, [ROW_FIELDS]);

  return (…);
};

export default CertificateTable;

File: src/tests/unit/pages/my-certificates/my-certificates.spec.js [hunk 13]
```diff
@@ -0,0 +1,95 @@
...
+);
+
+jest.mock(
+  '../../../../containers/my-certificates/certificate-table/certificate-table.styles.js',
+  () => ({
+    useStyles: () => ({})
+  })
+);
+
+jest.mock('@apollo/client');
```
Comment (maarekua) [inline src/tests/unit/pages/my-certificates/my-certificates.spec.js:32]: If we have some requests in components u wanna test, it is a bad idea to mock Appolo client.

https://www.apollographql.com/docs/react/development-testing/testing/

**Review by maarekua — CHANGES_REQUESTED** [review]

File: src/tests/unit/pages/my-certificates/my-certificates.spec.js (hunk 13 above)
Comment (anemyro) [reply src/tests/unit/pages/my-certificates/my-certificates.spec.js:32]: - [ x] FIXED

Thanks a lot for the link. I've seen it before but didn't pay enough attention on it. 
After refactoring "my-certificates.spec" according to the article I could increase the coverage from 70% t0 100%. 

--------BEFORE------------
Screenshot 2022-01-31 at 21 06 21

----------AFTER-----------
Screenshot 2022-01-31 at 21 01 19

**Review by boris-giga — APPROVED** [review]

**Pull request merged**